packages/ra-core/src/dataProvider/withLifecycleCallbacks.ts

```diff
@@ -462,107 +373,88 @@ export const withLifecycleCallbacks = (
-            }
+            result = await applyCallbacks(resource, 'afterDeleteMany', result);
 
             return result;
         },
     };
 };
 
+export type ResourceCallbackUtils = {
+    resource: string;
+    dataProvider: DataProvider;
+};
+
+export type ResourceCallback<U> = {
+    (params: U, utils: ResourceCallbackUtils): Promise<U>;
+};
+
 export type ResourceCallbacks<T extends RaRecord = any> = {
     resource: string;
-    afterCreate?: (
-        result: CreateResult<T>,
-        dataProvider: DataProvider
-    ) => Promise<CreateResult<T>>;
-    afterDelete?: (
-        result: DeleteResult<T>,
-        dataProvider: DataProvider
-    ) => Promise<DeleteResult<T>>;
-    afterDeleteMany?: (
-        result: DeleteManyResult<T>,
-        dataProvider: DataProvider
-    ) => Promise<DeleteManyResult<T>>;
-    afterGetList?: (
-        result: GetListResult<T>,
-        dataProvider: DataProvider
-    ) => Promise<GetListResult<T>>;
-    afterGetMany?: (
-        result: GetManyResult<T>,
-        dataProvider: DataProvider
-    ) => Promise<GetManyResult<T>>;
-    afterGetManyReference?: (
-        result: GetManyReferenceResult<T>,
-        dataProvider: DataProvider
-    ) => Promise<GetManyReferenceResult<T>>;
-    afterGetOne?: (
-        result: GetOneResult<T>,
-        dataProvider: DataProvider
-    ) => Promise<GetOneResult<T>>;
-    afterUpdate?: (
-        result: UpdateResult<T>,
-        dataProvider: DataProvider
-    ) => Promise<UpdateResult<T>>;
-    afterUpdateMany?: (
-        result: UpdateManyResult<T>,
-        dataProvider: DataProvider
-    ) => Promise<UpdateManyResult<T>>;
-    beforeCreate?: (
-        params: CreateParams<T>,
-        dataProvider: DataProvider
-    ) => Promise<CreateParams<T>>;
-    beforeDelete?: (
-        params: DeleteParams<T>,
-        dataProvider: DataProvider
-    ) => Promise<DeleteParams<T>>;
-    beforeDeleteMany?: (
-        params: DeleteManyParams<T>,
-        dataProvider: DataProvider
-    ) => Promise<DeleteManyParams<T>>;
-    beforeGetList?: (
-        params: GetListParams,
-        dataProvider: DataProvider
-    ) => Promise<GetListParams>;
-    beforeGetMany?: (
-        params: GetManyParams,
-        dataProvider: DataProvider
-    ) => Promise<GetManyParams>;
-    beforeGetManyReference?: (
-        params: GetManyReferenceParams,
-        dataProvider: DataProvider
-    ) => Promise<GetManyReferenceParams>;
-    beforeGetOne?: (
-        params: GetOneParams<T>,
-        dataProvider: DataProvider
-    ) => Promise<GetOneParams<T>>;
-    beforeUpdate?: (
-        params: UpdateParams<T>,
-        dataProvider: DataProvider
-    ) => Promise<UpdateParams<T>>;
-    beforeUpdateMany?: (
-        params: UpdateManyParams<T>,
-        dataProvider: DataProvider
-    ) => Promise<UpdateManyParams<T>>;
+    afterCreate?:
+        | ResourceCallback<CreateResult<T>>[]
+        | ResourceCallback<CreateResult<T>>;
```

Comment:
You should create a type for this union, to avoid repetition

Reply:
changed